File: 2020/julia/AOC/src/day3.jl

module Day3
function count_trees(slope, data)
	pos = [1, 1]
	count = 0
	while pos[1] < size(data, 1)
		pos .+= slope 
		count += data[pos[1], mod1(pos[2], size(data, 2))] 
	end
	count
end
end


function day3_part1(;data=Parsers.day3())
	Day3.count_trees([1, 3], data)
end

function day3_part2(;data=Parsers.day3())
	mapreduce(*, [[1, 1], [1, 3], [1, 5], [1, 7], [2, 1]]) do slope
		Day3.count_trees(slope, data)
	end
end

File: 2020/fsharp/AOC/Day3.fs
module AOC.Day3

open System

let countTrees (slopeX: int, slopeY: int) (data: int [,]) = 
    let xs = Seq.toArray (seq {for n in 0 .. slopeX .. (Array2D.length1 data - 1) do n})
    let ys = seq {for n in 0 .. (Array.length xs - 1) do (n * slopeY) % Array2D.length2 data}
    Seq.zip xs ys
    |> Seq.sumBy (fun (x, y) -> data.[x, y])

    
let part1 data = 
    int64 (countTrees (1, 3) data)

let part2 data = 
    Seq.fold (fun accum slope -> accum * int64 (countTrees slope data)) 
        1L
        [ (1, 1); (1, 3); (1, 5); (1, 7); (2, 1) ]
